### PR TITLE
SAI proposal for icmp echo offload

### PR DESCRIPTION
**Pull request by chikkaiah-work**
This specification suggests:
- The idea of an ICMP echo object or session that can be transferred to HW/ASIC or other modules for faster link detection.
- The introduction of SAI APIs to outline ICMP echo session properties.
- The introduction of a switch-level notification API to monitor changes in ICMP echo session states.

### CODE REVIEW

**Comment by kcudnik** [comment]
please fix build errors

**Review by rck-innovium — CHANGES_REQUESTED** [review]

**Comment by kcudnik** [comment]
please fix build errors, your new file does not meet our standard, read errors to help you guide: https://dev.azure.com/mssonic/build/_build/results?buildId=457039&view=logs&j=83516c17-6666-5250-abde-63983ce72a49&t=1f6d5e21-f0ce-508d-13df-d1c50f7ea7b9&l=1146

**Comment by kcudnik** [comment]
you will probably need squash commits as well, since our validator checks one by one

**Comment by kcudnik** [comment]
try to build this locally on your machine and make pass then submit commit to github

**Comment by chikkaiah-work** [comment]
> try to build this locally on your machine and make pass then submit commit to github



> try to build this locally on your machine and make pass then submit commit to github
I was able to test it locally and fix all meta check and build errors. thank you!

File: inc/saiicmpecho.h [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    SAI_ICMP_ECHO_SESSION_ATTR_START, 
+
+    /** 
+     * @brief Hardware lookup valid 
+     * 
+     * @type bool 
+     * @flags CREATE_ONLY 
+     * @default true 
+     */ 
+    SAI_ICMP_ECHO_SESSION_ATTR_VIRTUAL_ROUTER = SAI_ICMP_ECHO_SESSION_ATTR_START,
```
[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply inc/saiicmpecho.h]: thanks for reviewing it. yes, it was a typo and I fixed it.

File: doc/SAI-Proposal-ICMP-ECHO.md [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+   +-+-+-+-+-
+```
+
+### 2.2.0 ICMP echo payload packet ###
+
+Besides the ICMP echo request and reply packets, additional information about the ICMP echo session is essential and is transmitted within the payload.
+
+***Session ID:*** To correlate probe and reply packets with a session, especially when the session is offloaded to hardware, a unique identifier is required. This distinct identifier is conveyed within the ICMP payload field.
+
+***Session Cookie:*** The value of the session cookie is communicated through the ICMP payload field. The offloading engine interprets this value to decide whether the packet should undergo processing by the ICMP echo session or not.
```
[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply doc/SAI-Proposal-ICMP-ECHO.md:59]: there is session level attribute (SAI_ICMP_ECHO_SESSION_ATTR_COOKIE) is added to do this.

File: doc/SAI-Proposal-ICMP-ECHO.md [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |     Data ...
+   +-+-+-+-+-
+```
+
+### 2.2.0 ICMP echo payload packet ###
+
+Besides the ICMP echo request and reply packets, additional information about the ICMP echo session is essential and is transmitted within the payload.
+
+***Session ID:*** To correlate probe and reply packets with a session, especially when the session is offloaded to hardware, a unique identifier is required. This distinct identifier is conveyed within the ICMP payload field.
```
[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply doc/SAI-Proposal-ICMP-ECHO.md:57]: No RFC. It is a MSFT proprietary format. its definition can be found in https://github.com/sonic-net/SONiC/blob/master/doc/dualtor/active_active_hld.md (section 3.3.1 Link Prober) 
The session ID is required because in the dual tor deployment virtual IPs are used. Ethernet source and destination addresses as well as the IP source and destination address are the same for packets originating from both ToR devices.

[the remark this answers is not in the document]
Comment (rck-innovium) [reply doc/SAI-Proposal-ICMP-ECHO.md:57]: Thanks for the reference.  Please clarify the below differences between the formats defined in the Dual TOR HLD and this SAI proposal.

1.  Sequence number is 8B in the SONiC HLD, whereas it is 4B in the SAI PR.
2.  SONiC HLD has a 8B GUID, SAI PR has a 8B Session-ID. Are these two different names for the same field?

[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply doc/SAI-Proposal-ICMP-ECHO.md]: 1) sequence number is 8B. fixed the doc. 
2) 8B Session-ID is same as 8B GUID of SONiC HLD. I fixed the doc by changing the name.

File: doc/SAI-Proposal-ICMP-ECHO.md [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |                           GUID(MSB)                           |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |                           GUID(LSB)                           |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |                                                               |
+   +-+-+-+-+-+-+-+-+-+-+-+-+ Sequence Number +-+-+-+-+-+-+-+-+-+-+-+
+   |                                                               |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |TLV Type (SENTINAL: 0xFF)  |   TLV Length (0)                  |
```
Comment (lolyu) [inline doc/SAI-Proposal-ICMP-ECHO.md:79]: the tvl_type is of 1 bytes and tlv_len is 2 bytes

File: doc/SAI-Proposal-ICMP-ECHO.md [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+   +-+-+-+-+-
+```
+
+### 2.2.0 ICMP echo payload packet ###
+
+Besides the ICMP echo request and reply packets, additional information about the ICMP echo session is essential and is transmitted within the payload.
+
+***GUID:*** Global unique identifier for the ICMP echo session to correlate probe and reply packets with a session, especially when the session is offloaded to hardware, a unique identifier is required. This distinct identifier is conveyed within the ICMP payload field.
+
+***Session Cookie:*** The value of the session cookie is communicated through the ICMP payload field. The offloading engine interprets this value to decide whether the packet should undergo processing by the ICMP echo session or not.
```
Comment (lolyu) [inline doc/SAI-Proposal-ICMP-ECHO.md:59]: all sessions should have the same cookie on the same device, right?

Comment (chikkaiah-work) [reply doc/SAI-Proposal-ICMP-ECHO.md:60]: Each session can share the same cookie, but if a device needs to communicate with multiple devices, each session can have its own unique cookie. The decision to assign cookies to sessions is implementation-specific, and the device can implement a check to enforce the desired behavior.

File: doc/SAI-Proposal-ICMP-ECHO.md (hunk 4 above)
Comment (chikkaiah-work) [reply doc/SAI-Proposal-ICMP-ECHO.md]: fixed

**Review by kcudnik — APPROVED** [review]

**Review by lolyu — APPROVED** [review]

File: doc/SAI-Proposal-ICMP-ECHO.md (hunk 4 above)
Comment (lolyu) [reply doc/SAI-Proposal-ICMP-ECHO.md]: resolved

File: doc/SAI-Proposal-ICMP-ECHO.md (hunk 5 above)
Comment (lolyu) [reply doc/SAI-Proposal-ICMP-ECHO.md:60]: resolved

**Comment by eddyk-nvidia** [comment]
Why can't BFD be used for the same functionality - testing of reacheability and connectivity ? BFD SAI API is already defined

**Comment by chikkaiah-work** [comment]
> Why can't BFD be used for the same functionality - testing of reacheability and connectivity ? BFD SAI API is already defined

This requirement pertains to the SONIC Dualtor (active-active) scenario, which employs an ICMP-based link prober. For further insights, refer to the documentation at: https://github.com/sonic-net/SONiC/blob/master/doc/dualtor/active_active_hld.md. In this scenario, both links present the same IP and MAC address to the server, a configuration that is incompatible with BFD (Bidirectional Forwarding Detection) due to its requirement for unique IP addresses for a single hop.

**Review by zjswhhh — APPROVED** [review]
LGTM

**Comment by chikkaiah-work** [comment]
> I have some basic questions on the feature itself and then suggestions for spec.
> 
> Questions:
> 
>     1. Is this between two peering links or across hops?
> 
>     2. How do you differentiate between ICMP echo response being dropped because of other network conditions instead of link down?
> 
>     3. ICMP packets are typically handled in slow path. For this to work the end points MUST have HW offload capability. Is this understanding correct?

1.  It can be either.
2. There's no distinction. In this context, "link down" encompasses both scenarios.
3. Yes, that's correct. However, it should function with other combinations (such as HW offload to SW module, or SW module to SW module implementations) as long as the configured intervals accommodate the expected transaction rate or slowness.

**Comment by jefftant** [comment]
> This specification suggests:
> 
> * The idea of an ICMP echo object or session that can be transferred to HW/ASIC or other modules for faster link detection.
> * The introduction of SAI APIs to outline ICMP echo session properties.
> * The introduction of a switch-level notification API to monitor changes in ICMP echo session states.

Hi Chikke,

At 100ms granularity, I don't see a reason for HW offload.
If we want to go faster (which we should), as Eddy suggested, BFD is a much better alternative. 

Your BFD statement above is incorrect.

RFC5880 states:
Since multiple BFD sessions may be running between two systems, there needs to be a mechanism for demultiplexing received BFD packets to the proper session.
Each system MUST choose an opaque discriminator value that identifies each session, and which MUST be unique among all BFD sessions on the system. The local discriminator is sent in the My Discriminator field in the BFD Control packet, and is echoed back in the Your Discriminator field of packets sent from the remote end.   

Specifically to single hop BFD (which is the case here) RFC5881 states:
On a point-to-point link, the source address of a BFD Control packetvMUST NOT be used to identify the session.  This means that thevinitial BFD packet MUST be accepted with any source address, and thatvsubsequent BFD packets MUST be demultiplexed solely by the YourvDiscriminator field (as is always the case)

**Comment by chikkaiah-work** [comment]
> > This specification suggests:
> > 
> > * The idea of an ICMP echo object or session that can be transferred to HW/ASIC or other modules for faster link detection.
> > * The introduction of SAI APIs to outline ICMP echo session properties.
> > * The introduction of a switch-level notification API to monitor changes in ICMP echo session states.
> 
> Hi Chikke,
> 
> At 100ms granularity, I don't see a reason for HW offload. If we want to go faster (which we should), as Eddy suggested, BFD is a much better alternative.
> 
> Your BFD statement above is incorrect.
> 
> RFC5880 states: Since multiple BFD sessions may be running between two systems, there needs to be a mechanism for demultiplexing received BFD packets to the proper session. Each system MUST choose an opaque discriminator value that identifies each session, and which MUST be unique among all BFD sessions on the system. The local discriminator is sent in the My Discriminator field in the BFD Control packet, and is echoed back in the Your Discriminator field of packets sent from the remote end.
> 
> Specifically to single hop BFD (which is the case here) RFC5881 states: On a point-to-point link, the source address of a BFD Control packetvMUST NOT be used to identify the session. This means that thevinitial BFD packet MUST be accepted with any source address, and thatvsubsequent BFD packets MUST be demultiplexed solely by the YourvDiscriminator field (as is always the case)

Yes, the requirement is to go much faster than 100ms. Original requirement is from SONIC dualtor use case. 

You are right IP addresses. Thanks Jeff. 

I didn’t explain clearly. I meant to mention the following of RFC 5881:

Initialization and Demultiplexing

   In this application, there will be only a single BFD session between
   two systems over a given interface (logical or physical) for a
   particular protocol.  The BFD session must be bound to this
   interface.  As such, both sides of a session MUST take the "Active"
   role (sending initial BFD Control packets with a zero value of Your
   Discriminator), and any BFD packet from the remote machine with a
   zero value of Your Discriminator MUST be associated with the session
   bound to the remote system, interface, and protocol.

For example, in case of SONIC dualtor active-active use case, there will be two sessions between ToR and server over a single physical link. The server basically duplicates ICMP heartbeat to both ToRs, that means each tor will receive TWO icmp replies. One originates from local, the other is from peer.  That way, a ToR can monitor the health status of  server link and  its peer ToR as well. 

Couple of other reasons that I think this will help: 
-	Not all NF/devices support BFD as compared to ICMP.  
-	BFD has overhead of session establishment and negotiation. 

Please correct me If I’m wrong or  have missed something. 

Thanks
Chikke

**Comment by eddyk-nvidia** [comment]
> > > This specification suggests:
> > > 
> > > * The idea of an ICMP echo object or session that can be transferred to HW/ASIC or other modules for faster link detection.
> > > * The introduction of SAI APIs to outline ICMP echo session properties.
> > > * The introduction of a switch-level notification API to monitor changes in ICMP echo session states.
> > 
> > 
> > Hi Chikke,
> > At 100ms granularity, I don't see a reason for HW offload. If we want to go faster (which we should), as Eddy suggested, BFD is a much better alternative.
> > Your BFD statement above is incorrect.
> > RFC5880 states: Since multiple BFD sessions may be running between two systems, there needs to be a mechanism for demultiplexing received BFD packets to the proper session. Each system MUST choose an opaque discriminator value that identifies each session, and which MUST be unique among all BFD sessions on the system. The local discriminator is sent in the My Discriminator field in the BFD Control packet, and is echoed back in the Your Discriminator field of packets sent from the remote end.
> > Specifically to single hop BFD (which is the case here) RFC5881 states: On a point-to-point link, the source address of a BFD Control packetvMUST NOT be used to identify the session. This means that thevinitial BFD packet MUST be accepted with any source address, and thatvsubsequent BFD packets MUST be demultiplexed solely by the YourvDiscriminator field (as is always the case)
> 
> Hi Jeff, Yes, the requirement is to go much faster than 100ms. Original requirement is from SONIC dualtor use case.
> 
> You are right about IP addresses. Thank you correcting me.
> 
> I didn’t explain clearly. I meant to mention the following of RFC 5881: ...... Initialization and Demultiplexing
> 
> In this application, there will be only a single BFD session between two systems over a given interface (logical or physical) for a particular protocol. The BFD session must be bound to this interface. As such, both sides of a session MUST take the "Active" role (sending initial BFD Control packets with a zero value of Your Discriminator), and any BFD packet from the remote machine with a zero value of Your Discriminator MUST be associated with the session bound to the remote system, interface, and protocol. ...... For example, in case of SONIC dualtor active-active use case, there will be two sessions between ToR and server over a single physical link. The server basically duplicates ICMP heartbeat to both ToRs, that means each tor will receive TWO icmp replies. One originates from local, the other is from peer. That way, a ToR can monitor the health status of server link and its peer ToR as well.
> 
> Couple of other reasons that I think this will help:
> 
> * Not all NF/devices support BFD as compared to ICMP.
> * BFD has overhead of session establishment and negotiation.
> 
> Please correct me If I’m wrong or have missed something.
> 
> Thanks Chikke

Hi Chikke,

Even if you plan to change the default interval for ICMP probing in Dual-TOR from 100ms to something smaller, still the link is considered unhealthy after 3 retries. So, even if you configurre 100ms --> 50ms, still 50x3=150ms is big enough time that hardly justifies the HW offloading

**Comment by chikkaiah-work** [comment]
> > > > This specification suggests:
> > > > 
> > > > * The idea of an ICMP echo object or session that can be transferred to HW/ASIC or other modules for faster link detection.
> > > > * The introduction of SAI APIs to outline ICMP echo session properties.
> > > > * The introduction of a switch-level notification API to monitor changes in ICMP echo session states.
> > > 
> > > 
> > > Hi Chikke,
> > > At 100ms granularity, I don't see a reason for HW offload. If we want to go faster (which we should), as Eddy suggested, BFD is a much better alternative.
> > > Your BFD statement above is incorrect.
> > > RFC5880 states: Since multiple BFD sessions may be running between two systems, there needs to be a mechanism for demultiplexing received BFD packets to the proper session. Each system MUST choose an opaque discriminator value that identifies each session, and which MUST be unique among all BFD sessions on the system. The local discriminator is sent in the My Discriminator field in the BFD Control packet, and is echoed back in the Your Discriminator field of packets sent from the remote end.
> > > Specifically to single hop BFD (which is the case here) RFC5881 states: On a point-to-point link, the source address of a BFD Control packetvMUST NOT be used to identify the session. This means that thevinitial BFD packet MUST be accepted with any source address, and thatvsubsequent BFD packets MUST be demultiplexed solely by the YourvDiscriminator field (as is always the case)
> > 
> > 
> > Hi Jeff, Yes, the requirement is to go much faster than 100ms. Original requirement is from SONIC dualtor use case.
> > You are right about IP addresses. Thank you correcting me.
> > I didn’t explain clearly. I meant to mention the following of RFC 5881: ...... Initialization and Demultiplexing
> > In this application, there will be only a single BFD session between two systems over a given interface (logical or physical) for a particular protocol. The BFD session must be bound to this interface. As such, both sides of a session MUST take the "Active" role (sending initial BFD Control packets with a zero value of Your Discriminator), and any BFD packet from the remote machine with a zero value of Your Discriminator MUST be associated with the session bound to the remote system, interface, and protocol. ...... For example, in case of SONIC dualtor active-active use case, there will be two sessions between ToR and server over a single physical link. The server basically duplicates ICMP heartbeat to both ToRs, that means each tor will receive TWO icmp replies. One originates from local, the other is from peer. That way, a ToR can monitor the health status of server link and its peer ToR as well.
> > Couple of other reasons that I think this will help:
> > 
> > * Not all NF/devices support BFD as compared to ICMP.
> > * BFD has overhead of session establishment and negotiation.
> > 
> > Please correct me If I’m wrong or have missed something.
> > Thanks Chikke
> 
> Hi Chikke,
> 
> Even if you plan to change the default interval for ICMP probing in Dual-TOR from 100ms to something smaller, still the link is considered unhealthy after 3 retries. So, even if you configurre 100ms --> 50ms, still 50x3=150ms is big enough time that hardly justifies the HW offloading

Hi Eddy, 
With the HW offload, packet are exchanges at micro sec rate. Detection can be performed in a few miliseconds. Also, this greatly helps if switchover from active to standby done in the HW directly.

**Comment by chikkaiah-work** [comment]
> try to build this locally on your machine and make pass then submit commit to github

@kcudnik, I changed this proposal to be added as expermental. I have changed the PR accordingly. But, I'm getting the following error from the meta sanity check pipeline. can you please help me understand and how to fix it.
error: 
 ASSERT FAILED (on line 221): sai_switch_attr_t: values are not increasing by 1: last: 229 current: 231, should be marked as @flags?

thanks!

File: experimental/saiswitchextensions.h [hunk 6]
```diff
@@ -37,6 +37,17 @@ typedef enum _sai_switch_attr_extensions_t
...
+    /**
+     * @brief Set Switch ICMP_ECHO session state change event notification callback function passed to the adapter.
+     *
+     * Use sai_icmp_echo_session_state_change_notification_fn as notification function.
+     *
+     * @type sai_pointer_t sai_icmp_echo_session_state_change_notification_fn
+     * @flags CREATE_AND_SET
+     * @default NULL
+     */
+    SAI_SWITCH_ATTR_ICMP_ECHO_SESSION_STATE_CHANGE_NOTIFY,
```
Comment (kcudnik) [inline experimental/saiswitchextensions.h:49]: you should add  = SAI_SWITCH_ATTR_EXTENSIONS_RANGE_START

Comment (chikkaiah-work) [reply experimental/saiswitchextensions.h]: that fixed the issue. thank you!

**Review by kcudnik — APPROVED** [review]

**Comment by zjswhhh** [comment]
Hi @kcudnik - can you help merge this PR? Thanks!

**Comment by kcudnik** [comment]
Was. This already approved on SAI community meeting?

**Comment by chikkaiah-work** [comment]
> Was. This already approved on SAI community meeting?

yes, this was reviewed and approved to be considered it as extension feature.

**Comment by JaiOCP** [comment]
@kcudnik Please revert the merge.
Review is pending from other folks as well.

**Comment by kcudnik** [comment]
reverted,
@chikkaiah-work please be careful next time and not cause issues like this

**Comment by zjswhhh** [comment]
Hi @JaiOCP - IIRC it's agreed in the meeting that if we proceeded this as an extension, Broadcom wouldn't be against it. @chikkaiah-work has put the proposal under experimental path. 

My apologies for the miscommunication here. Please feel free to raise any other concerns you have in the PR discussion; we will surely address them before merging.

File: experimental/saiexperimentalicmpecho.h [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+    SAI_ICMP_ECHO_SESSION_ATTR_START,
+
+    /**
+     * @brief Hardware lookup valid
+     *
+     * @type bool
+     * @flags CREATE_ONLY
+     * @default true
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_HW_LOOKUP_VALID = SAI_ICMP_ECHO_SESSION_ATTR_START,
```
[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply experimental/saiexperimentalicmpecho.h:83]: yes, ICMP echo request packet is being routed by performing IP lookup.

File: experimental/saiexperimentalicmpecho.h [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+
+    /**
+     * @brief Destination Port
+     *
+     * @type sai_object_id_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_PORT
+     * @condition SAI_ICMP_ECHO_SESSION_ATTR_HW_LOOKUP_VALID == false
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_PORT,
```
[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply experimental/saiexperimentalicmpecho.h:103]: Yes, that's correct.

File: experimental/saiexperimentalicmpecho.h [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+
+    /**
+     * @brief Session Global Unique Identifier
+     *
+     * @type sai_uint64_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_GUID,
```
[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply experimental/saiexperimentalicmpecho.h:111]: it is a per session identifier. I can remove "Global" to avoid confusion.

File: experimental/saiexperimentalicmpecho.h [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+    SAI_ICMP_ECHO_SESSION_ATTR_SRC_MAC_ADDRESS,
+
+    /**
+     * @brief L2 destination MAC address
+     *
+     * @type sai_mac_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     * @condition SAI_ICMP_ECHO_SESSION_ATTR_HW_LOOKUP_VALID == false
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_DST_MAC_ADDRESS,
```
[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply experimental/saiexperimentalicmpecho.h:179]: I don't see mac addresses in sai_tam_collector_attr_t.

File: experimental/saiexperimentalicmpecho.h [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_TX_INTERVAL,
+
+    /**
+     * @brief Receive interval in microseconds
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_RX_INTERVAL,
```
[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply experimental/saiexperimentalicmpecho.h:195]: This is required to detect the session timeout, if the rx packets haven't received within the specified interval.

File: experimental/saiexperimentalicmpecho.h [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+
+    /**
+     * @brief To enable protection group switchover on session state change
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_SET_NEXT_HOP_GROUP_SWITCHOVER,
```
[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply experimental/saiexperimentalicmpecho.h:204]: protection group of type SAI_NEXT_HOP_GROUP_TYPE_PROTECTION is configured. In that member, 'ICMP echo session object' is specified in the SAI_NEXT_HOP_GROUP_MEMBER_ATTR_MONITORED_OBJECT. 
more details in https://github.com/opencomputeproject/SAI/blob/master/doc/SAI-Proposal-FRR.md

File: doc/SAI-Proposal-ICMP-ECHO_Extensions.md [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+
+
+## 1.0  Introduction
+
+The traditional use of ICMP-based link prober is to identify link failures. This prober sends ICMP echo packets, receives any ICMP echo replies, and reports the detection results.
+
+When control plane application carries out the link probe (software link failure detection), the system operates at a slow, periodic exchange of probe requests and replies since it uses the kernel network stack to send and receive control packets. This may prolong link failure detection. To decrease this detection interval, the application could utilize the HW/ASIC to send and receive link probe packets at a quicker rate.
+
+The objective of this document is to propose a SAI specification to simplify the ICMP-based probe as a basic service. The service primitives offered by ICMP echo include creating, destroying, and modifying a session, considering the destination address and other parameters. In response, ICMP Echo sends a signal to its clients when the ICMP Echo session starts or stops.
+
```
[the remark this answers is not in the document]
Comment (chikkaiah-work) [reply doc/SAI-Proposal-ICMP-ECHO_Extensions.md:20]: Sure, I can add some details about ICMP reachability consideration over multihop networks.

Regarding differentiating drop and actual link down. The proposal here is same as BFD.  
BFD is designed to be robust against isolated packet drops. Since it sends control packets at regular, short intervals, the loss of a single packet doesn't immediately trigger a link down event. Instead, the link is declared down only if a predetermined number of consecutive packets are missed. This threshold is configurable and can be set based on the expected level of packet loss in the network.

Similar to BFD, ICMP echo also has configurable RX & TX intervals. RX interval (threshold) can be set based on the expected level of packet loss in the network. 

However, it's important to note that BFD itself doesn't explicitly tell you whether a link is down due to a physical failure or due to packet drops; it only indicates that the expected BFD packets have not been received. 
So, in both cases, additional diagnostic steps or strategies may be required to determine the exact cause of the failure.

In summary, HW offloaded session (ASIC level) will just transmit packets periodically and report state of reception of icmp echo packets, as per the configured TX/RX intervals, to the application or client. Beyond this, application can implement, if required, other strategies to make sure its not false link failure.  

Please correct me, if I'm wrong or missed something.
